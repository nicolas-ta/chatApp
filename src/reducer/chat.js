export const chatReducer = (state, action) => {
  switch (action.type) {
    case 'refresh': {
      return {
        ...state,
        selectedChannel: undefined,
        messageMap: {},
        messages: [],
        loading: false,
        error: null,
      };
    }

    case 'switch-channel': {
      return {
        ...state,
        selectedChannel: action.payload,
      };
    }
    case 'fetch-messages': {
      const {messages} = action.payload || {};
      const distinctMessages = messages.filter(
        message => !state.messageMap[message.reqId],
      );
      const mergedMessages = [...state.messages, ...distinctMessages];
      for (let i = 0; i < mergedMessages.length - 1; i++) {
        mergedMessages[i].hasSameSenderAbove =
          mergedMessages[i].sender &&
          mergedMessages[i + 1].sender &&
          mergedMessages[i].sender.userId ===
            mergedMessages[i + 1].sender.userId;
      }

      const messageMap = {};
      for (let i in distinctMessages) {
        const message = distinctMessages[i];
        messageMap[message.reqId] = true;
      }
      return {
        ...state,
        messages: mergedMessages,
        messageMap,
        empty: mergedMessages.length === 0 ? 'Start conversation.' : '',
      };
    }
    case 'receive-message':
      const {message, clearInput} = action.payload || {};

      if (state.messages.length > 0) {
        message.hasSameSenderAbove =
          message.sender &&
          state.messages[0].sender &&
          message.sender.userId === state.messages[0].sender.userId;
      }

      return {
        ...state,
        messages: [message, ...state.messages],
        messageMap: {...state.messageMap, [message.reqId]: true},
        input: clearInput ? '' : state.input,
        empty: '',
      };
    case 'send-message': {
      const {message, clearInput} = action.payload || {};

      if (!state.messageMap[message.reqId]) {
        if (state.messages.length > 0) {
          message.hasSameSenderAbove =
            message.sender &&
            state.messages[0].sender &&
            message.sender.userId === state.messages[0].sender.userId;
        }

        return {
          ...state,
          messages: [message, ...state.messages],
          messageMap: {...state.messageMap, [message.reqId]: true},
          input: clearInput ? '' : state.input,
          empty: '',
        };
      } else {
        for (let i in state.messages) {
          if (state.messages[i].reqId === message.reqId) {
            const updatedMessages = [...state.messages];
            message.hasSameSenderAbove = updatedMessages[i].hasSameSenderAbove;
            updatedMessages[i] = message;

            return {
              ...state,
              input: clearInput ? '' : state.input,
              messages: updatedMessages,
            };
          }
        }
      }
      break;
    }

    case 'typing': {
      const {input} = action.payload || {};
      return {...state, input};
    }
    case 'start-loading': {
      return {...state, loading: true, error: ''};
    }
    case 'end-loading': {
      const {error} = action.payload || {};
      return {...state, loading: false, error};
    }
    case 'error': {
      const {error} = action.payload || {};
      return {...state, error};
    }
  }
  return state;
};
