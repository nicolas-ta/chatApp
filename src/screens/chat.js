import React, {useState} from 'react';
import {Box, Button} from 'native-base';
import {Animated} from 'react-native';
import chatStyle from '../styles/chatStyle';
import {VALUE} from '../misc/constants';

import ChannelSelection from '../components/chanelSelection';
import CurrentChat from '../components/currentChat';
import MemberList from '../components/memberList';

const Chat = props => {
  const {navigation, sendbird} = props;
  const [padding, setPadding] = useState(new Animated.Value(0));
  const [currentScreen, setCurrentScreen] = useState(1);

  /** Open the left panel showing the channels */
  const openMenu = () => {
    setCurrentScreen(0);
    Animated.timing(padding, {
      toValue: VALUE.chatPanOffset,
      duration: VALUE.chatPanDuration,
      useNativeDriver: false,
    }).start();
  };

  /** Open the right panel showing the connected members */
  const openOnlineMember = () => {
    setCurrentScreen(2);
    Animated.timing(padding, {
      toValue: -VALUE.chatPanOffset,
      duration: VALUE.chatPanDuration,
      useNativeDriver: false,
    }).start();
  };

  /** Return to the main chat screen */
  const reset = () => {
    Animated.timing(padding, {
      toValue: 0,
      duration: VALUE.chatPanDuration,
      useNativeDriver: false,
    }).start(() => {
      setCurrentScreen(1);
    });
  };

  /** Define the left panel */
  const channelPanel = (
    <ChannelSelection isCurrentScreen={currentScreen === 0} />
  );

  /** Define the main chat view */
  const currentChat = (
    <Animated.View style={[chatStyle.middleBox, {marginLeft: padding}]}>
      <CurrentChat
        openMenu={openMenu}
        openOnlineMember={openOnlineMember}
        overlay={
          currentScreen !== 1 ? (
            <Button onPress={reset} style={chatStyle.overlay} />
          ) : undefined
        }
      />
    </Animated.View>
  );

  /** Define the right panel */
  const memberPanel = <MemberList isCurrentScreen={currentScreen === 2} />;

  /** Define the background, behind the chat view, and in front of the left or right panel
   * depending on the selected screen
   */
  const bg = <Box style={chatStyle.background} />;

  return (
    <Box>
      {bg}
      {channelPanel}
      {memberPanel}
      {currentChat}
    </Box>
  );
};

export default Chat;
