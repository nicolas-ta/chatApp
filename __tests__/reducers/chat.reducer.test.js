import {chatReducer} from '@reducers/chat';

const mockMessage1 = {
  sender: {userId: 'userId1'},
  reqId: 'reqId1',
  hasSameSenderAbove: true,
};

const mockMessage2 = {
  sender: {userId: 'userId2'},
  reqId: 'reqId2',
  hasSameSenderAbove: false,
};

describe('Member Reducer unit tests', () => {
  test('checking initial state', () => {
    const action = {type: 'nonexistent'};
    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking refresh state', () => {
    const action = {type: 'refresh'};
    const state = {...state};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking fetch-messages empty state', () => {
    const action = {type: 'fetch-messages'};

    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking fetch-messages state', () => {
    const action = {type: 'fetch-messages', payload: {messages: []}};
    const state = {messageMap: [], messages: []};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking receive-message empty state', () => {
    const action = {type: 'receive-message'};
    const state = {messages: []};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking receive-message state', () => {
    const action = {type: 'receive-message', payload: {message: mockMessage1}};
    const state = {messages: []};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking send-message empty state', () => {
    const action = {
      type: 'send-message',
    };
    const state = {messageMap: [], messages: []};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking send-message state', () => {
    const action = {
      type: 'send-message',
      payload: {
        sentMsg: mockMessage1,
        sentClearInput: true,
      },
    };
    const state = {messageMap: {}, messages: []};
    expect(chatReducer(state, action)).toMatchSnapshot();
  });

  test('checking typing state', () => {
    const action = {type: 'typing'};
    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking start-loading state', () => {
    const action = {type: 'start-loading'};
    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking end-loading state', () => {
    const action = {type: 'end-loading'};
    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });

  test('checking error state', () => {
    const action = {type: 'error'};
    expect(chatReducer(undefined, action)).toMatchSnapshot();
  });
});
