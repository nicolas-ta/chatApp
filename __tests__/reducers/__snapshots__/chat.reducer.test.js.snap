// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member Reducer unit tests checking end-loading state 1`] = `
Object {
  "error": undefined,
  "loading": false,
}
`;

exports[`Member Reducer unit tests checking error state 1`] = `
Object {
  "error": undefined,
}
`;

exports[`Member Reducer unit tests checking fetch-messages empty state 1`] = `Object {}`;

exports[`Member Reducer unit tests checking fetch-messages state 1`] = `
Object {
  "empty": "",
  "messageMap": Object {
    "reqId1": true,
  },
  "messages": Array [
    Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
    Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId2",
      "sender": Object {
        "userId": "userId2",
      },
    },
    Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
  ],
}
`;

exports[`Member Reducer unit tests checking initial state 1`] = `undefined`;

exports[`Member Reducer unit tests checking receive-message empty state 1`] = `
Object {
  "empty": "",
  "input": undefined,
  "messageMap": Object {},
  "messages": Array [
    undefined,
  ],
}
`;

exports[`Member Reducer unit tests checking receive-message state 1`] = `
Object {
  "empty": "",
  "input": undefined,
  "messageMap": Object {
    "reqId1": true,
  },
  "messages": Array [
    Object {
      "hasSameSenderAbove": true,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
    Object {
      "hasSameSenderAbove": true,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
    Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId2",
      "sender": Object {
        "userId": "userId2",
      },
    },
  ],
}
`;

exports[`Member Reducer unit tests checking refresh state 1`] = `
Object {
  "error": null,
  "loading": false,
  "messageMap": Object {},
  "messages": Array [],
}
`;

exports[`Member Reducer unit tests checking send-message empty state 1`] = `
Object {
  "messageMap": Array [],
  "messages": Array [],
}
`;

exports[`Member Reducer unit tests checking send-message state 1`] = `
Object {
  "empty": "",
  "input": "",
  "messageMap": Object {
    "mockMessage2": Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId2",
      "sender": Object {
        "userId": "userId2",
      },
    },
    "reqId1": true,
  },
  "messages": Array [
    Object {
      "hasSameSenderAbove": true,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
    Object {
      "hasSameSenderAbove": true,
      "reqId": "reqId1",
      "sender": Object {
        "userId": "userId1",
      },
    },
    Object {
      "hasSameSenderAbove": false,
      "reqId": "reqId2",
      "sender": Object {
        "userId": "userId2",
      },
    },
  ],
}
`;

exports[`Member Reducer unit tests checking start-loading state 1`] = `
Object {
  "error": "",
  "loading": true,
}
`;

exports[`Member Reducer unit tests checking typing state 1`] = `
Object {
  "input": undefined,
}
`;
